src/iarc7_sensors/orientation_filter.py

```diff
@@ -0,0 +1,184 @@
+
+            for i in xrange(len(self._queue) - 1, -1, -1):
+                if self._queue[i][1].header.stamp < msg.header.stamp:
+                    index = i+1
+                    break
+            else:
+                index = 0
+
+            data = [float('NaN')] * 3
+            try:
+                data[1] = -msg.data.pitch
+                data[2] = msg.data.roll
+            except AttributeError:
+                # If this message type doesn't have pitch and roll, skip it
+                pass
+
+            self._queue.insert(index, (data, msg))
+            self._reprocess(index)
+
+            if (not math.isnan(self._queue[-1][0][0])
+                and not math.isnan(self._queue[-1][0][1])
+                and not math.isnan(self._queue[-1][0][2])):
+
+                new_stamp = max(self._last_published_stamp + rospy.Duration(0, 1),
+                                self._queue[-1][1].header.stamp)
+                self._publish_transform(*reversed(self._queue[-1][0]), time=new_stamp)
+                self._last_published_stamp = new_stamp
+
+            while len(self._queue) > self._message_queue_length:
+                self._queue.pop(0)
+
+    def _reprocess(self, index):
+        for i in xrange(index, len(self._queue)):
+            last_yaw = self._queue[i-1][0][0]
+            msg = self._queue[i][1]
+            last_orientation_i = self._get_last_index(OrientationAnglesStamped,
+                                                      msg.header.stamp)
+
+            if type(msg) == OrientationAnglesStamped:
+
+                if last_orientation_i == -1:
+                    dyaw = 0
+                else:
+                    dyaw = -(msg.data.yaw
+                           - self._queue[last_orientation_i][1].data.yaw)
+
+                if dyaw > math.pi:
+                    dyaw -= 2*math.pi
+                elif dyaw < -math.pi:
+                    dyaw += 2*math.pi
+
+                self._queue[i][0][0] = last_yaw + dyaw
+            else:
+                new_yaw = msg.data
+                if new_yaw - last_yaw > math.pi:
+                    new_yaw -= 2*math.pi
+                elif new_yaw - last_yaw < -math.pi:
+                    new_yaw += 2*math.pi
+
+                if math.isnan(last_yaw):
+                    self._queue[i][0][0] = new_yaw
+                else:
+                    self._queue[i][0][0] = (self._line_weight     * new_yaw
+                                          + (1-self._line_weight) * last_yaw)
+
+                if last_orientation_i != -1:
+                    self._queue[i][0][1] = self._queue[last_orientation_i][0][1]
+                    self._queue[i][0][2] = self._queue[last_orientation_i][0][2]
+
+            self._queue[i][0][0] += 2*math.pi
+            self._queue[i][0][0] %= 2*math.pi
+
+    def run(self):
+        self._safety_client = SafetyClient('orientation_filter')
+        assert self._safety_client.form_bond()
+
+        rate = rospy.Rate(100)
```

Comment:
There's really no need to have this thread waking up at 100Hz. Could we do 15Hz? The callbacks are all handled by other threads.